Sources/HTTP/HTTPHeadersAccessors.swift

```diff
@@ -0,0 +1,1677 @@
+            } else {
+                self.storage[.ifNoneMatch] = nil
+            }
+        }
+    }
+    
+    /// `If-Range` header etag value.
+    public var ifRange: HTTPHeaders.IfRange? {
+        get {
+            return self.storage[.ifRange]?.first.flatMap(IfRange.init)
+        }
+        set {
+            self.storage[.ifRange] = newValue.flatMap { [$0.description] }
+        }
+    }
+    
+    /// `If-Modified-Since` header value.
+    public var ifModifiedSince: Date? {
+        get {
+            return self.storage[.ifModifiedSince]?.first.flatMap(Date.init(rfcString:))
+        }
+        set {
+            self.storage[.ifModifiedSince] = newValue.flatMap { [$0.format(with: .http)] }
+        }
+    }
+    
+    /// `If-Unmodified-Since` header value.
+    public var ifUnmodifiedSince: Date? {
+        get {
+            return self.storage[.ifUnmodifiedSince]?.first.flatMap(Date.init(rfcString:))
+        }
+        set {
+            self.storage[.ifUnmodifiedSince] = newValue.flatMap { [$0.format(with: .http)] }
+        }
+    }
+    
+    /// `Origin` header value.
+    public var origin: URL? {
+        get {
+            return self.storage[.origin]?.first.flatMap(URL.init(string:))
+        }
+        set {
+            self.storage[.origin] = newValue.flatMap { [$0.absoluteString] }
+        }
+    }
+    
+    fileprivate func dissectRanges(_ value: String?) -> [(from: Int64, to: Int64?)] {
+        // Converting real negatives to _ to avoid conflict with - as separator
+        let value = value?.replacingOccurrences(of: "=-", with: "=_").replacingOccurrences(of: "--", with: "-_").replacingOccurrences(of: ",-", with: ",_")
+        guard let bytes = value?.components(separatedBy: "=").dropFirst().first?.trimmingCharacters(in: .whitespaces) else {
+            return []
+        }
+        
+        var ranges = [(from: Int64, to: Int64?)]()
+        for range in bytes.components(separatedBy: ",") {
+            let elements = range.components(separatedBy: "-").map({ $0.trimmingCharacters(in: .whitespaces) })
+            
+            guard let lower = elements.first.flatMap({ Int64($0.replacingOccurrences(of: "_", with: "-")) }) else {
+                continue
+            }
+            let upper = elements.dropFirst().first.flatMap { Int64($0.replacingOccurrences(of: "_", with: "-")) }
+            ranges.append((lower, upper))
+        }
+        return ranges
+    }
+    
+    /// Returns `Range` header values. An empty array means no value is set in header.
+    /// - Note: upperbound will be Int64.max in case of open ended Range.
+    /// - Note: Server response should be `multipart/byteranges` in case of more than one range is returned.
+    ///    See [MDN's HTTP range requests](https://developer.mozilla.org/en-US/docs/Web/HTTP/Range_requests) for more info.
+    /// - Important: A negative range means server must return last bytes/items of file.
```

Comment:
`bytes` is the most common type and to validate ranges one would get the file size from a file as: `(fileAttributes[FileAttributeKey.size] as? NSNumber)?.uint64Value` because file size are declared as `UInt64` in Foundation. However current declaration is `Int64` which means there would be cases when the string can not be converted to Range<Int64>. How can we address this?. 

Example: try to get a tiny part of a very big file: 
`Range: bytes=18446744073709551600-18446744073709551601`

Are you using `Int64` is because of ranges like: `Range: bytes=-3` which look like negative?. Actually there is no such a thing as negative ranges. In fact they should be converted into a  range object/struct **after** the file size is known, for example if fileSize is `1000` then it would mean: `Range: bytes=997-999` which uses only positive numbers.

I like the idea of having everything statically defined as structs/classes/enums rather than raw strings, but I am not sure having this interface will help to correctly implement the RFC :(

(As a side note, nodejs implementation converts range header string into range objects in a function that receives the fileSize as a parameter.)

Reply:
About Foundation: `HTTPURLResponse.expectedContentLength` and `URLSession` methods return `Int64`.

On negative values, We don't define how to interpret data as we don't know about resource. It's upon higher level framework to convert negative values to meaningful ones or return 416.

About BigInt, Using Decimal will hit performance. We are implementing these method for common cases. A file with that size is not common and higher level framework would have to implement theirs. Should we add limitations to documentation?

Reply:
OK, if other values are Int64 then I think this should be ok :]
IMO I don't think documentation needs to know about this. Other framework implementers will likely read comments in code